formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoBuf.kt

```diff
@@ -340,48 +372,93 @@ public class ProtoBuf(
             return -1
         }
 
-        override fun beginStructure(descriptor: SerialDescriptor, vararg typeParams: KSerializer<*>): CompositeDecoder =
-            when (descriptor.kind) {
-                StructureKind.LIST -> RepeatedReader(decoder, currentTag, descriptor)
-                StructureKind.CLASS, StructureKind.OBJECT, is PolymorphicKind ->
-                    ProtobufReader(makeDelimited(decoder, currentTagOrNull), descriptor)
-                StructureKind.MAP -> MapEntryReader(makeDelimited(decoder, currentTagOrNull), currentTagOrNull, descriptor)
+        override fun beginStructure(descriptor: SerialDescriptor, vararg typeParams: KSerializer<*>): CompositeDecoder {
+            return when (descriptor.kind) {
+                StructureKind.LIST -> RepeatedDecoder(reader, currentTagOrDefault, descriptor)
+                StructureKind.CLASS, StructureKind.OBJECT, is PolymorphicKind -> {
+                    val tag = currentTagOrNull
+                    // Do not create redundant copy
+                    if (tag == null && this.descriptor == descriptor) return this
```

Comment:
It advances it only if it's part of an inner object. `tag == null` checks that this is exactly not the case.
In fact, this check avoids cache populating twice because previously it first happened in `load` method (when creating a ProtobufDecoder) and then second time right after in `beginObject`.